[docs] Dev tools plugins docs (#26031)
expo/expo: docs/pages/debugging/create-devtools-plugins.mdx

---
title: Create a dev tools plugin
description: Learn how to create a dev tools plugin to enhance your development experience.
---

import { Terminal } from '~/ui/components/Snippet';

> **warning** Dev Tools plugins are compatible with SDK 50 and above. The process for creating a plugin is still in beta and may change in the future.

> You create a dev tools plugin, whether that's for inspecting aspects of a common framework or library or something specific to your custom code.

Reviewer: ```suggestion
You can create a dev tools plugin, whether that's for inspecting aspects of a common framework or library or something specific to your custom code. This guide will walk you through creating a dev tools plugin.
```

Add the context as to what this guide does for me as a reader (and why I should care about it).

A plugin consists of three key elements:

> - an Expo app to display the dev tools web user interface
> - an **expo-module.config.json** for Expo CLI recognition
> - calls to `expo/devtools` API for the app to communicate back and forth with the dev tool's web interface

Reviewer 2: ```suggestion
- an Expo app to display the dev tools web user interface,
- an **expo-module.config.json** for Expo CLI recognition,
- calls to `expo/devtools` API for the app to communicate back and forth with the dev tool's web interface.
```

Formatting nit. Alternatively, we can make the points a full statements (start them with upper case), and then we should end the lines with period.

> A plugin consists of three key elements:
> - an Expo app to display the dev tools web user interface,
> - an **expo-module.config.json** for Expo CLI recognition,
> - calls to `expo/devtools` API for the app to communicate back and forth with the dev tool's web interface.

Reviewer: ```suggestion
A plugin consists of three key elements:
- An Expo app to display the dev tools web user interface.
- An **expo-module.config.json** for Expo CLI recognition.
- Calls to `expo/devtools` API for the app to communicate back and forth with the dev tool's web interface.
```

We generally avoid the pattern of lowercase and using `,` for list items. Let's separate them into different sentences for consistency.

This feels like it could be its own section (separate from the introduction).

+  "react-dom": "18.2.0",
+  "react-native": "0.73.0",
+  "react-native-web": "~0.19.6"
},
```

Add a `homepage` field to define the exported base URL for web in the format of `/_expo/plugins/{pluginName}`. For example:

```diff package.json
  "name": "expo-plugin-helloworld",
  "version": "1.0.0",
+  "homepage": "/_expo/plugins/expo-plugin-helloworld",
```

### Add message sending and receiving code to the plugin's web UI

`getDevToolsPluginClientAsync` initiates a client that can send messages and listen for incoming messages:

```js
  import { getDevToolsPluginClientAsync } from 'expo/devtools';

  const client = await getDevToolsPluginClientAsync('REPLACE-FOR-YOUR-PLUGIN-NAME');

  client.sendMessage('ping');
  client.addMessageListener('pong', (data) => {
    console.log('receiving pong', data);
  });
```

A `useDevToolsPluginClient` hook is also available to provide a client entirely within a React component.

Client apps using your plugin will also be able to communicate using `getDevToolsPluginClientAsync` or `useDevToolsPluginClient`. The identifier passed to `getDevToolsPluginClientAsync` must match in both the plugin and the client app.

### Build the app for web

> Create a **package.json** script that builds the web app and moves it to the **dist** directory. This will be served by the Expo CLI when invoking the plugin.

Reviewer: ```suggestion
Create a **package.json** script to build the web app and move it to the **dist** directory. This will be served by the Expo CLI when invoking the plugin.
```

```diff package.json
"scripts": {
  "start": "expo start",
  "android": "expo start --android",
  "ios": "expo start --ios",
  "web": "expo start --web",
+  "build": "rm -rf dist web-build && expo export:web && mv web-build dist"
},
```

> Then run the build script to export the plugin web UI and move it to the **dist** folder:

Reviewer: ```suggestion
Run the build script to export the plugin web UI and move it to the **dist** folder:
```

import { APIBox } from '~/components/plugins/APIBox';

> **warning** Dev Tools plugins are compatible with SDK 50 and above.

You can create a dev tools plugin, whether that's for inspecting aspects of a common framework or library or something specific to your custom code. This guide will walk you through creating a dev tools plugin.

## What is a dev tools plugin?
A dev tools plugin runs in your web browser in your local development environment and connects to your Expo app.

A plugin consists of three key elements:
- An Expo app to display the dev tools web user interface.
- An **expo-module.config.json** for Expo CLI recognition.
- Calls to `expo/devtools` API for the app to communicate back and forth with the dev tool's web interface.

Plugins can be distributed on npm or included inside your app's monorepo. They typically export a single hook that can be used in your app's root component to initiate two-way communication with the web interface when your app is running in debug mode.

## Create a plugin

### Create a new plugin project

`create-dev-plugin` will setup a new plugin project for you. Run the following command to create a new plugin project:

> <Terminal cmd={['$ npx create-dev-plugin']} />

Reviewer 3: ```suggestion
<Terminal cmd={['$ npx create-dev-plugin@latest']} />
```

recommend to use the latest dist-tag for npx, in case npx cache the package and keep using the old version

`create-dev-plugin` will prompt you for a name of your plugin, a description, and the name of the hook that will be used by consumers of your plugin.

The plugin project will contain the following folders:
- **src** - this exports the hook that will be used inside the consuming app to connect it to the plugin.
- **webui** - this contains the web user interface for the plugin.

> ### Customize your plugin's functionality

Reviewer: ```suggestion
### Customize a plugin's functionality
```

  const client = useDevToolsPluginClient('my-devtools-plugin');

  useEffect(() => {
    const subscriptions: EventSubscription[] = [];

    subscriptions.push(
      client?.addMessageListener('ping', (data) => {
        alert(`Received ping from ${data.from}`);
      })
    );

    return () => {
      for (const subscription of subscriptions) {
        subscription?.remove();
      }
    };
  }, [client]);
}

```

Edit the hook in the **src** folder to customize what diagnostic information is sent to the plugin or how the app should respond to any messages from the web user interface:
```tsx src/useMyDevToolsPlugin.ts
import { useDevToolsPluginClient } from 'expo/devtools';

export function useMyDevToolsPlugin() {
  const client = useDevToolsPluginClient('my-devtools-plugin');

  const sendPing = () => {
    client?.sendMessage('ping', { from: 'app' });
  }

  return {
    sendPing,
  }
}
```

If you update the hook to return functions that will be called by the app, you will also need to update **src/index.ts** so it exports no-op functions when the app is not running in debug mode:
```diff src/index.ts
if (process.env.NODE_ENV !== 'production') {

>   useMyDevToolsPlugin = require('./useMyDevToolsPlugin'). useMyDevToolsPlugin;

Reviewer 3: ```suggestion
  useMyDevToolsPlugin = require('./useMyDevToolsPlugin').useMyDevToolsPlugin;
```

my fault to have an extra space

>   useTestPlugin = require('./useTestPlugin').useTestPlugin;
> } else {
>   useTestPlugin = () => ({

Reviewer 3: ```suggestion
  useMyDevToolsPlugin = require('./useMyDevToolsPlugin'). useMyDevToolsPlugin;
} else {
  useMyDevToolsPlugin = () => ({
```

inconsistent hook name

+    sendPing: () => {},
  });
}
```

> ### Test your plugin

Reviewer: ```suggestion
### Test a plugin
```

Since the plugin web UI is an Expo app, you can test it just like you would any other Expo app, with `npx expo start`, except that you will run it in the browser only. The template includes a convenience command to run the plugin in local development mode:

<Terminal cmd={['$ npm run web:dev']} />

> ### Build your plugin

Reviewer: ```suggestion
### Build a plugin for distribution
```

To prepare your plugin for distribution or for use within your monorepo, you will need to build the plugin with the following command:

> <Terminal cmd={['$ npm run prepublishOnly']} />

Reviewer 3: ```suggestion
<Terminal cmd={['$ npm run build:all']} />
```

i changed my mind for the command. even though prepublishOnly still work, but the semantic looks strange. i've added a `build:all` command for this.

This command will build the hook code into the **build** folder, and the web user interface into the **dist** folder.

### Use your plugin
Import the plugin's hook into your app's root component and call it to connect your app to the plugin:

```tsx App.js
import { useMyDevToolsPlugin } from 'my-devtools-plugin';
import { Button } from 'react-native';

export default function App() {

>   const { sendPing } = useTestPlugin();

Reviewer 3: ```suggestion
  const { sendPing } = useMyDevToolsPlugin();
```

inconsistent hook name
